### PR TITLE
Restructure and update differential rover module

### PR DESCRIPTION
**Pull request by chfriedrich98**
### Solved Problem
This PR adresses 2 issues:
1. Fixes naming inconsistency in the rover related modules
2. Restructures differential module and improves guidance by implementing the pure pursuit algorithm (https://github.com/PX4/PX4-Autopilot/pull/23387).

### Solution
1. Rename module from differential drive to rover differential (same naming convention as rover ackermann module)
2. Restructure:

2a. Bring folder structure in line with ackermann module:
image
2b. With the restructure in 2a the `DifferentialDriveSetpoint` message becomes obsolete and is instead replaced with `RoverDifferentialStatus` and `RoverDifferentialGuidanceStatus` which are used exclusively for logging purposes (in line with the ackermann module):
`RoverDifferentialStatus` (General logging):

| Variable | Description | Unit |
|--------|--------|--------|
| actual_speed | Actual speed of the rover | m/s |
| desired_yaw_rate | Desired yaw rate of the rover | deg/s |
| actual_yaw_rate | Actual yaw rate of the rover | deg/s |
| pid_yaw_rate_integral | Integral of the PID for the desired yaw rate controller | - |

`RoverDifferentialGuidanceStatus` (Mission specific logging):

| Variable | Description | Unit |
|--------|--------|--------|
| desired_speed | Desired speed of the rover | m/s |
| lookahead_distance | Lookahead distance of pure the pursuit controller | m |
| heading_error | Heading error of the pure pursuit controller | deg |
| pid_heading_integral | Integral of the PID for the desired yaw rate during missions | - |
| pid_throttle_integral | Integral of the PID for the throttle during missions | - |

2c. Implement pure pursuit in the guidance algorithm:
The differential rover guidance is now based on the same algorithm as the ackermann rover (https://github.com/PX4/PX4-Autopilot/pull/23387). The rover will now actually track the line segment between waypoints, rather than just target the next waypoint, leading to a lower overall crosstrack error.
To fully exploit the 'turn on the spot' capabilities of differential rovers (as opposed to ackermann) the following state machine is implemented:
image

This causes the rover to stop and turn on the spot if the heading error exceeds a certain threshhold.
This decreases the accumulated crosstrack error, especially whenever the waypoint triplet changes.

### Alternatives
Open to any suggestions

### Test coverage
- SITL tested

https://github.com/user-attachments/assets/f76198c4-10b0-4f47-ae9c-bfe9d853d142

- Hardware tested (Vehicle: [Aion robotics R1](https://resources.basicmicro.com/aion-robotics-r1-autonomous-robot/))
pr_log
[Flight Log](https://review.px4.io/plot_app?log=09992e91-22b1-4a41-9917-f618c26f8b18).

### CODE REVIEW

File: src/modules/rover_differential/RoverDifferential.cpp [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+		_differential_setpoint.yaw_rate = 0.f;
+		_differential_setpoint.closed_loop_yaw_rate = false;
+		break;
+	}
+
+	float speed_diff_normalized = _differential_setpoint.yaw_rate;
+
+	// Closed loop yaw rate control
+	if (_differential_setpoint.closed_loop_yaw_rate) {
+		if (fabsf(_differential_setpoint.yaw_rate - _vehicle_body_yaw_rate) < 0.1f) {
```
Comment (sfuhrer) [inline src/modules/rover_differential/RoverDifferential.cpp:158]: can you add comment for the 0.1 and/or make it a constant?

File: src/modules/rover_differential/module.yaml [hunk 2]
```diff
@@ -120,3 +112,59 @@ parameters:
...
+        unit: m/s
+        min: 0
+        max: 100
+        increment: 0.01
+        decimal: 2
+        default: 7
+
+      RD_MAX_YAW_RATE:
+        description:
+          short: Maximum yaw rate for the rover.
```
Comment (sfuhrer) [inline src/modules/rover_differential/module.yaml:130]: This description isn't clear enough.

File: src/modules/rover_differential/module.yaml [hunk 3]
```diff
@@ -120,3 +112,59 @@ parameters:
         increment: 0.01
         decimal: 2
         default: 0.5
+
+      RD_MAX_SPEED:
+        description:
+          short: Maximum speed
+          long: Maximum speed the rover can drive.
```
Comment (sfuhrer) [inline src/modules/rover_differential/module.yaml:119]: Is this intended to be used as setpoint limit? Then I would specify it here.

File: src/modules/rover_differential/module.yaml [hunk 4]
```diff
@@ -120,3 +112,59 @@ parameters:
...
+        unit: deg/s
+        min: 0.01
+        max: 1000
+        increment: 0.01
+        decimal: 2
+        default: 90
+
+      RD_MISS_VEL_DEF:
+        description:
+          short: Default rover velocity during a mission
```
Comment (sfuhrer) [inline src/modules/rover_differential/module.yaml:141]: It's a speed and not a velocity, no?

File: src/modules/rover_differential/module.yaml [hunk 5]
```diff
@@ -120,3 +112,59 @@ parameters:
...
+            short: Heading error threshhold to switch from spot turning to driving
+        type: float
+        unit: rad
+        min: 0.001
+        max: 180
+        increment: 0.01
+        decimal: 3
+        default: 0.0872665
+
+      RD_TRANS_DRV_TRN:
```
Comment (sfuhrer) [inline src/modules/rover_differential/module.yaml:161]: I wonder if those really need to be two different params. Could you also work with a hysteresis but with the same threshold for both directions?

File: msg/RoverDifferentialStatus.msg [hunk 6]
```diff
@@ -0,0 +1,8 @@
+uint64 timestamp # time since system start (microseconds)
+
+float32 actual_speed	      # [m/s] Actual forward speed of the rover
+float32 desired_yaw_rate      # [deg/s] Desired yaw rate
```
Comment (sfuhrer) [inline msg/RoverDifferentialStatus.msg:4]: and here maybe add `_deg_s`

**Review by sfuhrer — COMMENTED** [review]
So we don't want to keep the inverse kinematics? I guess most rovers don't do wheel speed control anyway and don't care about it.
Looks very clean otherwise!

File: src/modules/rover_differential/module.yaml (hunk 5 above)
Comment (chfriedrich98) [reply src/modules/rover_differential/module.yaml:161]: Not dismissing this as a valid option, but I usually tuned it s.t. the transition to go from driving to turning is more than double the transition from turning to driving. My idea is that you only want to come to a stop and do a spot turn if the heading error is quite large but once you come to a stop you should take the time to get the orientation as good as possible before continuing, therfore needing a lower threshold.

**Comment by chfriedrich98** [comment]
@sfuhrer 

> So we don't want to keep the inverse kinematics? I guess most rovers don't do wheel speed control anyway and don't care about it.

Correct me if I'm wrong but the inverse kinematics of the old implementation didn't do direct wheel speed control either: https://github.com/PX4/PX4-Autopilot/blob/b93dd0e8d45340e510e74b5be84a618b15a95787/src/modules/differential_drive/DifferentialDriveKinematics/DifferentialDriveKinematics.cpp#L70-L97

I implemented the same inverse kinematics, but in a "normalized" form to directly compute the motor inputs (yaw rate setpoint is already transformed into a normalized required speed_diff before going into this function) : https://github.com/PX4/PX4-Autopilot/blob/817bd21793f15b456cec5164999401bd3de6e1e6/src/modules/rover_differential/RoverDifferential.cpp#L193-L205

File: src/modules/rover_differential/module.yaml (hunk 3 above)
Comment (chfriedrich98) [reply src/modules/rover_differential/module.yaml]: Specified it more in 817bd21

File: src/modules/rover_differential/module.yaml (hunk 2 above)
Comment (chfriedrich98) [reply src/modules/rover_differential/module.yaml]: Updated in 817bd21

File: src/modules/rover_differential/RoverDifferential.cpp (hunk 1 above)
Comment (chfriedrich98) [reply src/modules/rover_differential/RoverDifferential.cpp]: implemented in 817bd21

**Comment by chfriedrich98** [comment]
rebased on main and squashed

File: src/modules/rover_differential/RoverDifferential.cpp [hunk 7]
```diff
@@ -0,0 +1,256 @@
...
+{
+	float combined_velocity = fabsf(forward_speed) + fabsf(speed_diff);
+
+	if (combined_velocity > 1.0f) { // Prioritize yaw rate
+		float excess_velocity = fabsf(combined_velocity - 1.0f);
+		forward_speed -= sign(forward_speed) * excess_velocity;
+	}
+
+	// Calculate the left and right wheel speeds
+	return Vector2f(forward_speed - speed_diff,
```
Comment (sfuhrer) [inline src/modules/rover_differential/RoverDifferential.cpp:203]: Shouldn't a positive yaw rate setpoint (speed_diff) result in an increase on the left and a reduction on the right?

**Review by sfuhrer — APPROVED** [review]
Agree that we can simplify the inverse kinematics like you propose. If we ever have somebody with a rover driver that takes wheel speed as input (and not throttle) we can revisit it.

File: src/modules/rover_differential/RoverDifferential.cpp (hunk 7 above)
Comment (sfuhrer) [reply src/modules/rover_differential/RoverDifferential.cpp:203]: (I assume ActuatorMotors(_left_, _right_)

Comment (chfriedrich98) [reply src/modules/rover_differential/RoverDifferential.cpp:203]: It seems that the definition is the other way around ActutorMotors(right, left)
image
It also works as intended on hardware tests.

**Comment by slgrobotics** [comment]
I tried ```"make px4_sitl gz_lawnmower_lawn"```, and it completes the mission all right. With the inertia and friction of a simulated big zero-turn lawnmower I should've adjusted default parameters to achieve smooth motion.

Overall, it's nice to see a clean implementation, I am tempted to move some of it to my custom code and see it working on my Husquarna Z254F. I am wondering if switching from L1 to Pure Pursuit will achieve better line following. Thanks for you work!

**Comment by chfriedrich98** [comment]
@slgrobotics Thanks for testing the changes!
> I tried "make px4_sitl gz_lawnmower_lawn", and it completes the mission all right. With the inertia and friction of a simulated big zero-turn lawnmower I should've adjusted default parameters to achieve smooth motion.

For the r1 rover I added my tuning directly to the SITL airframe s.t. it works "out of the box" when you start the simulation, which I think is what people would expect. I should do the same for the landmower, thanks for the reminder!

> Overall, it's nice to see a clean implementation, I am tempted to move some of it to my custom code and see it working on my Husquarna Z254F. I am wondering if switching from L1 to Pure Pursuit will achieve better line following. Thanks for you work!

Thank you! Please keep me updated if you use it on your rover, it's always very insightful to see how it works on different vehicles.
From my testing I found the pure pursuit algorithm to work a lot better than l1, but of course it still comes down to proper tuning.

**Comment by slgrobotics** [comment]
@chfriedrich98  Please see https://github.com/PX4/PX4-Autopilot/pull/23527

I am sure there's room for perfecting/tuning these parameters further, but for now the vehicle provides reasonable demo - even when running a realistic lawn mowing "survey" with 1 meter distance between the stripes.

**Comment by slgrobotics** [comment]
@chfriedrich98 @dagar  - please take a look at https://github.com/PX4/PX4-Autopilot/pull/23611  (Stanley Pursuit implementation offered)

**Pull request merged**